File: 1.rb
# https://leetcode.com/problems/two-sum/


# O(n)
def two_sum(nums, target)
	res = []
	hash = {}
	nums.each_with_index { |num, i|
		missing = target - num
		if hash[missing]
			res << i << hash[missing]
		else
			hash[num] = i
		end
	}
	res
end

# O(n2)
def two_sum(nums, target)
	sum = []
	nums.each_with_index { |num1, i|
		nums.each_with_index { |num2, j|
			next if i == j

			sum << i << j if num1 + num2 == target
		}
	}
	sum.uniq
end

File: js/1.js
// https://leetcode.com/problems/two-sum/submissions/
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
	if (!nums || !nums.length) {
		return []
	} else if (nums.length === 1 && nums[0] === target) {
		return [0]
	} else if (nums.length === 2) {
		if (nums[0] + nums[1] === target) {
			return [0,1]
		} else {
			return []
		}
	}
	let i = 0, hash = {}
	while (i<nums.length) {
		if (hash[nums[i]] !== undefined) {
			return [hash[nums[i]], i]
		} else {
			hash[target-nums[i]] = i
		}
		i++
	}
};
